### PR TITLE
Add matchPattern syntax to check if the result matches certain patterns

### PR DESCRIPTION
**Pull request by Kevin-Lee**
Add matchPattern syntax to check if the result matches certain patterns

### CODE REVIEW

**Comment by Kevin-Lee** [comment]
I had several test cases at my work requiring this kind of pattern matching capability for assertion.

File: core/shared/src/main/scala/hedgehog/package.scala [hunk 1]
```diff
@@ -31,12 +31,15 @@ package object hedgehog extends ApplicativeSyntax {
...
 
-  implicit class Syntax[A](a1: A) {
+  implicit class Syntax[A](private val a1: A) extends AnyVal {
 
     // FIX Is there a way to get this to work with PropertyT and type-inference?
     def ====(a2: A): Result =
       Result.diffNamed("=== Not Equal ===", a1, a2)(_ == _)
 
-}
+    def matchPattern(right: PartialFunction[A, _]): Result =
```
Comment (charleso) [inline core/shared/src/main/scala/hedgehog/package.scala:40]: The only suggestion/tweak I might be suggested to make is `PartialFunction[A, Result]` (possibly as a separate method).

Comment (charleso) [inline core/shared/src/main/scala/hedgehog/package.scala:40]: I really wish there was a super no-dependency, light-weight library _just_ for `Result`, this kind of matching is really it's own space where everyone has an opinion.

**Review by charleso — APPROVED** [review]

File: core/shared/src/main/scala/hedgehog/package.scala (hunk 1 above)
Comment (Kevin-Lee) [reply core/shared/src/main/scala/hedgehog/package.scala:40]: I think that can be already achieved by just
```scala
actual match {
  case Something(1, _) => Result.success
  case _ => Result.failure
}
```
This PR is more about convenience.

Comment (Kevin-Lee) [reply core/shared/src/main/scala/hedgehog/package.scala:40]: Yeah I agree. I'm not so sure about this PR either as everyone may have an option just like you said.
So I just put something similar to what ScalaTest has https://www.scalatest.org/user_guide/using_matchers#matchingAPattern

Do you think it should be provided as a third-party extension? 🤔

Comment (charleso) [reply core/shared/src/main/scala/hedgehog/package.scala:40]: copying ScalaTest sounds good/fine to me. Just ignore me :)

**Pull request merged**